### PR TITLE
zero config

### PR DESCRIPTION
**Pull request by bblay**
First version of zero config command line tool.

Usage:
```
$ fab
$ fab <folder>
```

`automatically found the following programs to build: second, first`

Tested on:
 - some system test projects
    - FortranDependencies
    - CFortranInterop

So far, only tested using gfortran. That needs a linker flag, `-lgfortran`. We should try it with ifort, too. There might be a few linker flags we'll want to add to cover a wide variety of use cases.

todo:
 - [ ]  test it with other stuff
 - [ ]  test it with ifort

### CODE REVIEW

File: source/fab/cli.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# ##############################################################################
+#  (c) Crown copyright Met Office. All rights reserved.
+#  For further details please refer to the file COPYRIGHT
+#  which you should have received as part of this distribution
+# ##############################################################################
+from argparse import ArgumentParser
+from pathlib import Path
+
+from fab.steps.link import LinkExe
```
Comment (bblay) [inline source/fab/cli.py:9]: pycharm keeps doing this with auto imports 😠

File: source/fab/cli.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from fab.steps.root_inc_files import RootIncFiles
+
+
+def _generic_build_config(folder: Path, kwargs: Optional[Dict] = None) -> BuildConfig:
+    folder = folder.resolve()
+    kwargs = kwargs or {}
+
+    # Within the fab workspace, we'll create a project workspace.
+    # Ideally we'd just use folder.name, but to avoid clashes, we'll use the full absolute path.
+    label = '/'.join(folder.parts[1:])
```
Comment (MatthewHambley) [inline source/fab/cli.py:31]: Is there  a way to do this using only Path objects so as to avoid string handling? Is it important to do that if the resulting label is needed as a string anyway?

File: source/fab/cli.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    fpp, fpp_flags = get_fortran_preprocessor()
+    fc, fc_flags = get_fortran_compiler()
+
+    if fc == 'gfortran':
+        link_step = LinkExe(linker='gcc', flags=['-lgfortran'])
+    else:
+        raise NotImplementedError(f"Fab's zero config not yet configured for compiler: '{fc}'")
```
Comment (MatthewHambley) [inline source/fab/cli.py:39]: In my "mesh tools" script I made use of a dictionary to easily map compiler to argument list. Would that approach make sense here? e.g.

```python
fortran_linker = {
    'gfortran': ('gcc, ['-lgfortran']),
    'ifort': (..., [...])
}
...
try:
    link_step = LineExe(linker=fortran_link[fc][0], flags=fortran_link[fc][1])
except KeyError:
    raise NotImplementedError(f"Fab's zero configuration mode does not yet work with compiler '{fc}'"
```

File: source/fab/cli.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        steps=[
+            GrabFolder(folder),
+            FindSourceFiles(),
+
+            RootIncFiles(),  # JULES helper, get rid of this eventually
+
+            fortran_preprocessor(preprocessor=fpp, common_flags=fpp_flags),
+
+            CPragmaInjector(),
+            c_preprocessor(source=CollectionGetter(PRAGMAD_C)),
```
Comment (MatthewHambley) [inline source/fab/cli.py:52]: The zero configuration build mode should definitely look for C files. When it gains the ability to output a sample configuration script it should only include C (or Fortran) if they were used in the build.

File: source/fab/cli.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    return config
+
+
+def cli_fab():
+    """
+    Running Fab from the command line will attempt to build the project in the current or given folder.
+
+    """
+    arg_parser = ArgumentParser()
+    arg_parser.add_argument('folder', nargs='?', default='.', type=Path)
```
Comment (MatthewHambley) [inline source/fab/cli.py:73]: Setting the default to `Path()` will have the same effect. I don't know if it's any better, just thought I'd mention it.

File: source/fab/cli.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    arg_parser.add_argument('-v', '--version', action='store_true')
```
Comment (MatthewHambley) [inline source/fab/cli.py:74]: `argparse` provides a `version` action which means you don't have to handle this yourself.

File: source/fab/parse/fortran.py [hunk 7]
```diff
@@ -185,7 +198,7 @@ def walk_nodes(self, fpath, file_hash, node_tree) -> AnalysedFortran:
                         analysed_fortran.add_symbol_dep(called_name.string)
 
                 elif obj_type == Program_Stmt:
-                    analysed_fortran.add_symbol_def(str(obj.get_name()))
+                    analysed_fortran.add_program_def(str(obj.get_name()))
```
Comment (MatthewHambley) [inline source/fab/parse/fortran.py:201]: This change might also fix the "unused symbols" check incorrectly listing program symbols. See #219.

File: source/fab/steps/analyse.py [hunk 8]
```diff
@@ -167,6 +177,21 @@ def run(self, artefact_store: Dict, config):
...
+            # find fortran programs
+            sets_of_programs = [af.program_defs for af in by_type(analysed_files, AnalysedFortran)]
+            self.root_symbols = list(chain(*sets_of_programs))
+
+            # find c main()
+            c_with_main = list(filter(lambda c: 'main' in c.symbol_defs, by_type(analysed_files, AnalysedC)))
+            if c_with_main:
+                self.root_symbols.append('main')
+                if len(c_with_main) > 1:
+                    raise FabException("multiple c main() functions found")
```
Comment (MatthewHambley) [inline source/fab/steps/analyse.py:191]: It's okay for there to be multiple C programs defined by the source tree. If you don't want to handle that with this change then raise an issue or project card for it.

File: source/fab/steps/analyse.py [hunk 9]
```diff
@@ -167,6 +177,21 @@ def run(self, artefact_store: Dict, config):
...
         self._add_manual_results(analysed_files)
 
+        # shall we search the results for fortran programs and a c function called main?
+        if self.find_programs:
+            # find fortran programs
+            sets_of_programs = [af.program_defs for af in by_type(analysed_files, AnalysedFortran)]
+            self.root_symbols = list(chain(*sets_of_programs))
+
+            # find c main()
+            c_with_main = list(filter(lambda c: 'main' in c.symbol_defs, by_type(analysed_files, AnalysedC)))
```
Comment (MatthewHambley) [inline source/fab/steps/analyse.py:187]: It may not be enough just to check for symbols called "main". I think we should be okay as class methods called "main" will get mangled but can you have a global variable called "main"? We don't necessarily need to handle these issues in this change but we should know what the possible failure modes are.

File: source/fab/steps/compile_fortran.py [hunk 10]
```diff
@@ -346,6 +346,42 @@ def compile_file(self, analysed_file, flags, output_fpath):
             value={'time_taken': timer.taken, 'start': timer.start})
 
 
+def get_fortran_preprocessor():
+    fpp, fpp_flags = None, None
+
+    try:
+        fpp, fpp_flags = get_tool(os.getenv('FPP'))
+        logger.info(f'env defined fpp as "{fpp}"')
```
Comment (MatthewHambley) [inline source/fab/steps/compile_fortran.py:354]: Shorthand is fine in debug messages but for "Info" level messages I think we prefer whole words and recognised contractions with apostrophes.

File: source/fab/steps/compile_fortran.py [hunk 11]
```diff
@@ -346,6 +346,42 @@ def compile_file(self, analysed_file, flags, output_fpath):
...
+    except ValueError:
+        pass
+
+    if not fpp:
+        try:
+            run_command(['which', 'fpp'])
+            fpp, fpp_flags = 'fpp', ['-P']
```
Comment (MatthewHambley) [inline source/fab/steps/compile_fortran.py:361]: A note about the purpose of "-P" and a link to https://github.com/stfc/fparser/issues/390

File: source/fab/steps/compile_fortran.py [hunk 12]
```diff
@@ -346,6 +346,42 @@ def compile_file(self, analysed_file, flags, output_fpath):
...
+        try:
+            run_command(['which', 'cpp'])
+            fpp, fpp_flags = 'cpp', ['-traditional-cpp', '-P']
+            logger.info('detected cpp')
+        except RuntimeError:
+            # fpp not available
+            pass
+
+    if not fpp:
+        raise RuntimeError('no fortran preprocessor specified or discovered')
```
Comment (MatthewHambley) [inline source/fab/steps/compile_fortran.py:377]: In the future we might consider a slicker implementation which takes a list of candidates and iterates until one succeeds.

File: source/fab/steps/compile_fortran.py [hunk 13]
```diff
@@ -356,8 +392,38 @@ def get_fortran_compiler(compiler: Optional[str] = None):
...
+            pass
+
+    if not fortran_compiler:
+        try:
+            run_command(['ifort', '--help'])
+            fortran_compiler = 'ifort', []
+            logger.info('detected ifort')
+        except RuntimeError:
+            # gfortran not available
+            pass
```
Comment (MatthewHambley) [inline source/fab/steps/compile_fortran.py:421]: The preference order for compilers is tricky. GFortran is faster to compile and more widely deployed but it bit rubbish. IFort is slower in compile and less common but produces much better executables. Some thought needs to be given this.

The fact that we go to the `FC` environment variable may be all the solution we need.

File: system-tests-new/zero_config/test_zero_config.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+class TestZeroConfig(object):
+
+    def test_fortran_dependencies(self, tmp_path):
+        # test the sample project in the fortran dependencies system test
+        config = _generic_build_config(
+            Path(__file__).parent.parent / 'FortranDependencies',
+            kwargs={'fab_workspace': tmp_path, 'multiprocessing': False})
+        config.run()
+        assert (config.project_workspace / 'first.exe').exists()
+        assert (config.project_workspace / 'second.exe').exists()
```
Comment (MatthewHambley) [inline system-tests-new/zero_config/test_zero_config.py:15]: Is it enough to test for the existence of an executable or do we need to run it as well?

File: tests/unit_tests/steps/test_compile_fortran.py [hunk 15]
```diff
@@ -457,3 +458,71 @@ def test_vanilla(self):
...
+
+        assert fpp == 'foo_pp'
+        assert fpp_flags == ['--foo', '-P']
+
+    def test_empty_env_fpp(self):
+        def mock_run_command(command):
+            if 'fpp' not in command:
+                raise RuntimeError('foo')
+
+        with mock.patch.dict(os.environ, values={'FPP': ''}):
```
Comment (MatthewHambley) [inline tests/unit_tests/steps/test_compile_fortran.py:477]: Does this remove the `FPP` variable from the environment  or simple set it to the empty string?

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
I'm excited to see movement on this mode of operation. This looks like a good starting point for further development.

File: source/fab/cli.py (hunk 2 above)
Comment (bblay) [reply source/fab/cli.py:31]: We _could_ change `label` to be a path, as it's mainly used to create a subfolder in the fab workspace. I'm not sure we should bother doing that, so yes, it's probably not important, as you say.

File: source/fab/cli.py (hunk 3 above)
Comment (bblay) [reply source/fab/cli.py:39]: Yes, that makes perfect sense. It's the natural progression as we add more compilers but we might as well do that right away.

File: source/fab/cli.py (hunk 6 above)
Comment (bblay) [reply source/fab/cli.py:74]: ooh lovely, thanks.

File: source/fab/parse/fortran.py (hunk 7 above)
Comment (bblay) [reply source/fab/parse/fortran.py:201]: Sadly not, as it calls `add_symbol_def` itself. Should still be an easy fix when we come to it though.

File: source/fab/steps/analyse.py (hunk 8 above)
Comment (bblay) [reply source/fab/steps/analyse.py:191]: :+1: made [a ticket ](https://github.com/orgs/metomi/projects/1/views/1?pane=issue&itemId=21047392)

File: source/fab/steps/analyse.py (hunk 9 above)
Comment (bblay) [reply source/fab/steps/analyse.py:187]: Agreed, although that feels like an extreme edge case.

File: system-tests-new/zero_config/test_zero_config.py (hunk 14 above)
Comment (bblay) [reply system-tests-new/zero_config/test_zero_config.py:15]: I think we should run all the exes we produce in our system tests, but I'm not sure it's a high priority yet.

File: tests/unit_tests/steps/test_compile_fortran.py (hunk 15 above)
Comment (bblay) [reply tests/unit_tests/steps/test_compile_fortran.py:477]: It sets it to an empty string.

**Comment by bblay** [comment]
Just noticed a todo I'd put at the top, to test with ifort. Will do that.

File: source/fab/steps/compile_fortran.py (hunk 11 above)
Comment (MatthewHambley) [reply source/fab/steps/compile_fortran.py:361]: Is GitHub playing silly buggers? you've acknowledged this comment but I don't see a comment in the diff.

File: tests/unit_tests/steps/test_compile_fortran.py [hunk 16]
```diff
@@ -457,3 +458,71 @@ def test_vanilla(self):
...
+
+        assert fpp == 'fpp'
+        assert fpp_flags == ['-P']
+
+    def test_empty_env_cpp(self):
+        def mock_run_command(command):
+            if 'cpp' not in command:
+                raise RuntimeError('foo')
+
+        with mock.patch.dict(os.environ, values={'FPP': ''}):
```
Comment (MatthewHambley) [inline tests/unit_tests/steps/test_compile_fortran.py:489]: Another case of setting `FPP` to the empty string.

File: tests/unit_tests/steps/test_compile_fortran.py [hunk 17]
```diff
@@ -457,3 +458,71 @@ def test_vanilla(self):
...
+
+        assert fc == 'foo_c'
+        assert fc_flags == ['--foo']
+
+    def test_empty_env_gfortran(self):
+        def mock_run_command(command):
+            if 'gfortran' not in command:
+                raise RuntimeError('foo')
+
+        with mock.patch.dict(os.environ, values={'FC': ''}):
```
Comment (MatthewHambley) [inline tests/unit_tests/steps/test_compile_fortran.py:511]: And another.

File: tests/unit_tests/steps/test_compile_fortran.py [hunk 18]
```diff
@@ -457,3 +458,71 @@ def test_vanilla(self):
...
+
+        assert fc == 'gfortran'
+        assert fc_flags == []
+
+    def test_empty_env_ifort(self):
+        def mock_run_command(command):
+            if 'ifort' not in command:
+                raise RuntimeError('foo')
+
+        with mock.patch.dict(os.environ, values={'FC': ''}):
```
Comment (MatthewHambley) [inline tests/unit_tests/steps/test_compile_fortran.py:523]: More empty strings

**Review by MatthewHambley — CHANGES_REQUESTED** [review]
Just a few more bits of tidying up.

File: source/fab/steps/compile_fortran.py (hunk 11 above)
Comment (bblay) [reply source/fab/steps/compile_fortran.py:361]: I think you'll find it is _I_ who is the silly bugger. Done now.

File: tests/unit_tests/steps/test_compile_fortran.py (hunk 16 above)
Comment (bblay) [reply tests/unit_tests/steps/test_compile_fortran.py:489]: :+1: thank you

File: tests/unit_tests/steps/test_compile_fortran.py (hunk 17 above)
Comment (bblay) [reply tests/unit_tests/steps/test_compile_fortran.py:511]: gosh

**Review by MatthewHambley — APPROVED** [review]
Time to bring this to trunk.
